src/Forms/Shared/CategoryListPage.xaml.cs

```diff
@@ -1,40 +1,143 @@
 //
-// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an 
-// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific 
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
 // language governing permissions and limitations under the License.
 
-using ArcGISRuntimeXamarin.Managers;
+using System;
+using ArcGISRuntime.Samples.Managers;
+using ArcGISRuntime.Samples.Shared.Models;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Esri.ArcGISRuntime.Security;
 using Xamarin.Forms;
 
-namespace ArcGISRuntimeXamarin
+namespace ArcGISRuntime
 {
-    public partial class CategoryListPage : ContentPage
+    public partial class CategoryListPage
     {
-        List<TreeItem> _sampleCategories;
-        public CategoryListPage ()
+        public CategoryListPage()
         {
             Initialize();
-            InitializeComponent ();
+            InitializeComponent();
         }
 
-        async void Initialize()
+        private void Initialize()
         {
-            await SampleManager.Current.InitializeAsync();
-            _sampleCategories= SampleManager.Current.GetSamplesAsTree();
-            BindingContext = _sampleCategories;
+            // Initialize the sample manager.
+            SampleManager.Current.Initialize();
 
+            ViewModel = new SamplesSearchViewModel();
+
+            // Update the binding.
+            BindingContext = ViewModel;
+        }
+
+        private async void OnItemTapped(object sender, ItemTappedEventArgs e)
+        {
+            // Get the selected category.
+            var category = e.Item as SearchableTreeNode;
+            var sample = e.Item as SampleInfo;
+            if (category != null)
+            {
+                // Navigate to the listing page for the category.
+                await Navigation.PushAsync(new SampleListPage(category.Name));
+            }
+            else if (sample != null)
+            {
+                // Navigate to the sample page
+                ShowSample(sample);
+            }
+        }
+
+        private async void ShowSample(SampleInfo item)
+        {
+            try
+            {
+                // Load offline data before showing the sample.
+                if (item.OfflineDataItems != null)
+                {
+                    // Show the wait page.
+                    await Navigation.PushAsync(new WaitPage { Title = item.SampleName }, false);
+
+                    // Wait for the sample data download.
+                    await DataManager.EnsureSampleDataPresent(item);
```

Comment:
It's possible for user to navigate away from the WaitPage while we're awaiting this. This leads to some unexpected behavior when the rest of the method resumes. I recommend checking to see if we're still _on the same WaitPage_ before executing the rest of the method. I'd also double-check EnsureSampleDataPresent for any potential race conditions, since it's possible that it may be running multiple times concurrently. While you're at it, you can avoid flashing CategoryListPage after the download by using `Navigation.RemovePage` in place of `PopAsync`.

...Alternatively, you could use PushModalAsync/PopModalAsync to prevent user from navigating away as the wait page is displayed.

Reply:
In testing, RemovePage (instead of PopAsync) still results in the flashed page on Android. This doesn't seem to be avoidable. I tried using modals; using remove page; only removing the page after the sample has been shown, etc.

PushModalAsync also doesn't prevent user navigation. On Android it changes the UI (hides the back button), but Android devices have a physical/virtual back button that still works.